extern/sector-storage/selector_alloc.go

```diff
@@ -64,8 +64,50 @@ func (s *allocSelector) Ok(ctx context.Context, task sealtasks.TaskType, spt abi
 	return false, nil
 }
 
-func (s *allocSelector) Cmp(ctx context.Context, task sealtasks.TaskType, a, b *workerHandle) (bool, error) {
-	return a.utilization() < b.utilization(), nil
+func (s *allocSelector) Cmp(ctx context.Context, task sealtasks.TaskType, spt abi.RegisteredSealProof, a, b *workerHandle) (bool, error) {
+	if s.ptype == storiface.PathSealing {
+		return a.utilization() < b.utilization(), nil
+	}
+	return s.storageCmp(ctx, spt, a, b)
+}
+
+func (s *allocSelector) storageCmp(ctx context.Context, spt abi.RegisteredSealProof, a, b *workerHandle) (bool, error) {
+	aPaths, err := a.workerRpc.Paths(ctx)
+	if err != nil {
+		return false, xerrors.Errorf("getting worker paths: %w", err)
+	}
+	aHave := map[stores.ID]struct{}{}
+	for _, path := range aPaths {
+		aHave[path.ID] = struct{}{}
+	}
+
+	bPaths, err := b.workerRpc.Paths(ctx)
+	if err != nil {
+		return true, xerrors.Errorf("getting worker paths: %w", err)
+	}
+	bHave := map[stores.ID]struct{}{}
+	for _, path := range bPaths {
+		bHave[path.ID] = struct{}{}
+	}
+
+	ssize, err := spt.SectorSize()
+	if err != nil {
+		return true, xerrors.Errorf("getting sector size: %w", err)
+	}
+	best, err := s.index.StorageBestAlloc(ctx, s.alloc, ssize, s.ptype)
```

Comment:
Calling `StorageBestAlloc` is too expensive to do in `Cmp`, at the very least we should cache that call